Show all annotations as top-level cards in streamer

gergely-ujvari opened the pull request:
First step for streamer and viewer separation.
StreamSearchController generates the cards to be displayed
in the streamer, and handles the annotation events.

Stream no longer relies on the threading service.
The messageContainers from jwz are still used.
They still share the same template.

Replies in the stream are displayed as root level annotation
cards.

Fix #1916

gergely-ujvari commented:
Third version of #2061 
Second version of #2069 

The cards class is not a service, because we do not use it anywhere else

tilgovi commented:
It's fine, but a little OO for my taste.

Can we keep it super simple and just inline the callbacks in the controller? The `cards` root container never changes, it can just be a local variable.

gergely-ujvari commented:
Okay, deOOPed, handled the login/logout case and added tests for StreamSearchController.

Ready for review.

For the further separation I'll open a separate issue in our vision repo

gergely-ujvari commented:
Actually I opened https://github.com/hypothesis/h/issues/2089 for it, and that can be linked to https://github.com/hypothesis/vision/issues/166

gergely-ujvari commented:
Actually it needs a bit more work.

gergely-ujvari commented:
Okay. Now this is really done.
I moved the threading service from AppController, added hook functions for looking up annotations, updated the tests.

nickstenning commented:
Rebased this for you.

nickstenning commented on h/static/scripts/test/annotation-viewer-controller-test.coffee:
```diff
@@ -5,17 +5,81 @@ sinon.assert.expose assert, prefix: null
 
 describe 'AnnotationViewerController', ->
-  annotationViewerController = null
+  $scope = null
```
Duplicate line.

nickstenning commented on h/static/scripts/test/streamsearch-test.coffee:
```diff
@@ -0,0 +1,182 @@
+    angular.module('h', ['ngRoute'])
+    require('../streamsearch')
+
+  describe 'StreamSearchController', ->
```
I think this must be my fault... going to look into what happened to the rebase here.

gergely-ujvari commented:
@nickstenning: Thank you very much for this!

gergely-ujvari commented:
Can I get a reviewer for this PR?

tilgovi commented:
The only general uncertainty I have about this is that we're using the scope to register different implementations depending on our controller. We know that a separation of the templates might be a good idea, but part of that is really about making the `AppController` also not responsible for too much. I wonder if we want to move the `streamer.onmessage` assignment inside the view controllers and just let each view handle updates in its own way.

gergely-ujvari commented:
@tilgovi: What about the `cleanupAnnotations()` inside the AppController? That is responsible for deleting annotations clientside.
Should any template have this power? (If yes, than they cannot really coexists)

tilgovi commented:
Template? No.

I think that whole function could move, though.

I don't like that we try to preserve cards across auth changes. It's a
performance hack, and it's a little risky (if something goes wrong we could
have non-public data still after sign-out).

Anyway, if the route reloads on auth change then it has to reconstruct its
view anyway. If annotations are cleaned up in the meantime, then everything
should work.

If this is too much work to rethink on this PR, we could merge this first,
but I'd like to see us at least think about what it would cost to improve.
On Apr 1, 2015 6:10 AM, "gergely-ujvari" <notifications@github.com> wrote:

> @tilgovi <https://github.com/tilgovi>: What about the cleanupAnnotations()
> inside the AppController? That is responsible for deleting annotations
> clientside.
> Should any template have this power? (If yes, than they cannot really
> coexists)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/2081#issuecomment-88473968>.
>

tilgovi commented on h/static/scripts/stream-controller.coffee:
```diff
@@ -1,17 +1,70 @@
+    $scope.lookup.getAnnotationContainers = ->
+      $scope.threadRoot.children
+
+    $rootScope.$on 'beforeAnnotationCreated', (event, annotation) =>
```
These will need to be unbound on `$destroy`

gergely-ujvari commented:
:+1: just for cleaning every template and don't try to be smart.
I'll check this.

tilgovi commented:
The trickiest things to watch out for are that we don't lose unsaved highlights when signing in. Right now, those get instantly persisted.

gergely-ujvari commented:
@tilgovi: I've created a branch to play here: https://github.com/hypothesis/h/tree/separate-controllers
It has no support, no fixed tests, but moved everything out from the AppController and it is much cleaner.
Do you support that way? (If yes, I'll make the rest)

tilgovi commented:
I think I like it and I suspect it'll get simpler before we're done.

gergely-ujvari commented:
Closing in favor of #2125
